Add Darwin arm64 cross compile script

joetoddsonos opened the pull request:
It should be possible to cross compile for `arm64` architectures like so. 

```
./mac_build.sh arm64
```

joetoddsonos committed:
Add Darwin arm64 cross compile script

bastibe commented:
Exciting! Thank you for trying this!

joetoddsonos committed:
Update github runner to build macos arm64

bastibe commented:
Please let me know when you once you need feedback or help.

joetoddsonos commented:
I have hit a bit of a blocker with this since I rebased and included MP3/Lame support. The check lame libs section of the configure script fails to validate the library: 

```
checking for library containing lame_set_VBR_q... no
configure: WARNING: "MPEG support selected but external Lame library cannot be found.
configure: WARNING: *** MPEG support disabled.
```

As far as I can tell the lame library does include this symbol, so I'm a bit stuck. 

```
➜  lame-3.100 git:(feature/macos-xcorr-build) ✗ lipo -info libmp3lame/.libs/libmp3lame.a
Non-fat file: libmp3lame/.libs/libmp3lame.a is architecture: arm64
➜  lame-3.100 git:(feature/macos-xcorr-build) ✗ strings libmp3lame/.libs/libmp3lame.a | grep lame_set_VBR_q
_lame_set_VBR_q
_lame_set_VBR_quality
```

I will slowly chip away at this if it doesn't block the 0.11 release 👍

Reference: https://github.com/bastibe/libsndfile-binaries/runs/7840829162?check_suite_focus=true#step:3:6055.

bastibe commented:
That sounds like an error in libsndfile's configure script using the wrong platform configuration for this check. If so, we should probably open an issue at libsndfile and ask for help there.

> I will slowly chip away at this if it doesn't block the 0.11 release 👍

Perfect! I think we have all the binaries we need for the 0.11 release for the moment. The present pull request will be important for whenever we want to ship an updated version of libsndfile, but that's probably many months in the future.

joetoddsonos commented:
Just making a note here that the searching for the lame library breaks because I am setting CFLAGS, so need to find another way of setting the target/arch.

bastibe commented:
Is there any progress on this? It would really help me a lot!

(and even if not, do you have an ARM mac available to build a recent version of libsndfile 1.2.0 on?)

joetoddsonos committed:
Try build libsndfile 1.2.0

joetoddsonos committed:
Try build libsndfile 1.2.0

joetoddsonos commented:
I will try rerun the build script, this might have been fixed in v1.2.0 by https://github.com/libsndfile/libsndfile/pull/822/files

joetoddsonos commented:
That last build looked like it worked! 🎉 No longer getting the warnings about libmp3lame. 
I can ask someone with an ARM machine to test it out.

```
➜  Downloads lipo -info libsndfile_arm64.dylib
Non-fat file: libsndfile_arm64.dylib is architecture: arm64
➜  Downloads otool -L libsndfile.dylib
libsndfile.dylib:
	/usr/local/lib/libsndfile.1.dylib (compatibility version 2.0.0, current version 2.35.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
```

bastibe commented:
That's amazing, thank you so much!

There is one warning `configure: WARNING: using cross tools not prefixed with host triplet`, do you know what that means?

Also, it seems that it compiles fine in arm64 mode, but then proceeds to link in x86_64 mode, and obviously fails. I think this is because we are missing the `CFLAGS=$EXTRA_CFLAGS CXXFLAGS=$EXTRA_CFLAGS` on the `configure` line of libsndfile.

bastibe commented on mac_build.sh:
```diff
@@ -80,17 +90,22 @@ export VORBISENC_CFLAGS="-I$(pwd)/libvorbis-$VORBISVERSION/include"
 tar jxvf libsndfile-$SNDFILE_VERSION.tar.xz
 cd $SNDFILENAME
-./configure --disable-static --disable-sqlite --disable-alsa
+./configure $BUILD_HOST --disable-static --disable-sqlite --disable-alsa --disable-full-suite
```
This is missing `CFLAGS=$EXTRA_CFLAGS CXXFLAGS=$EXTRA_CFLAGS`, I think.

joetoddsonos replied:
I think when you add these CFLAGS back, it fails to find the `lame` library again

joetoddsonos commented:
Sorry I deleted that last comment because I got ahead of myself. The output tricked me because the actions just uploaded the library that already exists in the repo, not the one that was built. 

It looks like I am still getting the same error, so I have changed tack to try and use cmake for building libsndfile. This does succeed in locating all of the locally built libraries. However I am now stuck at a different error:

```
ld: warning: ignoring file /Users/joe.todd/Repos/hacks/libsndfile-binaries/flac-1.4.2/src/libFLAC/libFLAC.la, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x23 0x20 0x6C 0x69 0x62 0x46 0x4C 0x41 0x43 0x2E 0x6C 0x61 0x20 0x2D 0x20 0x61 )
ld: warning: ignoring file /Users/joe.todd/Repos/hacks/libsndfile-binaries/opus-1.3.1/libopus.la, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x23 0x20 0x6C 0x69 0x62 0x6F 0x70 0x75 0x73 0x2E 0x6C 0x61 0x20 0x2D 0x20 0x61 )
ld: warning: ignoring file /Users/joe.todd/Repos/hacks/libsndfile-binaries/libogg-1.3.5/src/libogg.la, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x23 0x20 0x6C 0x69 0x62 0x6F 0x67 0x67 0x2E 0x6C 0x61 0x20 0x2D 0x20 0x61 0x20 )
ld: warning: ignoring file /Users/joe.todd/Repos/hacks/libsndfile-binaries/mpg123-1.30.2/src/libmpg123/libmpg123.la, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x23 0x20 0x6C 0x69 0x62 0x6D 0x70 0x67 0x31 0x32 0x33 0x2E 0x6C 0x61 0x20 0x2D )
ld: warning: ignoring file /Users/joe.todd/Repos/hacks/libsndfile-binaries/lame-3.100/libmp3lame/.libs/libmp3lame.la, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x23 0x20 0x6C 0x69 0x62 0x6D 0x70 0x33 0x6C 0x61 0x6D 0x65 0x2E 0x6C 0x61 0x20 )
ld: warning: ignoring file /Users/joe.todd/Repos/hacks/libsndfile-binaries/libvorbis-1.3.7/lib/libvorbis.la, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x23 0x20 0x6C 0x69 0x62 0x76 0x6F 0x72 0x62 0x69 0x73 0x2E 0x6C 0x61 0x20 0x2D )
ld: warning: ignoring file /Users/joe.todd/Repos/hacks/libsndfile-binaries/libvorbis-1.3.7/lib/libvorbis.la, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x23 0x20 0x6C 0x69 0x62 0x76 0x6F 0x72 0x62 0x69 0x73 0x2E 0x6C 0x61 0x20 0x2D )
```

I have checked with lipo and these libraries are built for arm64, so no idea what this unsupported file format means?.

bastibe commented:
Interesting! On the CI, the same error message reads [`building for macOS-x86_64 but attempting to link with file built for macOS-arm64`](https://github.com/bastibe/libsndfile-binaries/actions/runs/4076864812/jobs/7035569178#step:3:6279).

bastibe commented:
The latest run seems to have worked 🎉

joetoddsonos commented:
Yeah that is a confusing message, lipo disagrees with its output :)

Unfortunately the lame linking did not work, even though the build succeeded. 

```
checking for library containing lame_set_VBR_q... no
configure: WARNING: "MPEG support selected but external Lame library cannot be found.
checking for libmpg123 >= 1.25.10 ... yes
configure: WARNING: *** MPEG support disabled.
....
External MPEG Lame/MPG123 : ........... no
```

I think the cmake build may be the best way forward, do feel like it's close to working, might be able to spend some more time on it today.

bastibe commented on mac_build.sh:
```diff
@@ -80,17 +90,22 @@ export VORBISENC_CFLAGS="-I$(pwd)/libvorbis-$VORBISVERSION/include"
-export MPG123_CFLAGS="-I$(pwd)/mpg123-$MPG123VERSION/src/libmpg123 $LAME_CFLAGS"
-export MPG123_LIBS="$(pwd)/mpg123-$MPG123VERSION/src/libmpg123/libmpg123.la $LAME_LIBS"
+export MPEG_CFLAGS="-I$(pwd)/lame-$LAMEVERSION/include"
+export MPEG_LIBS="$(pwd)/lame-$LAMEVERSION/libmp3lame/.libs/libmp3lame.la"
```
According to [the build log](https://github.com/bastibe/libsndfile-binaries/actions/runs/4082021218/jobs/7035934334#step:3:5253), the output directory might be `libmp3lame/libmp3lame.la`. It looks like only the `.a` goes to `.libs`.

joetoddsonos replied:
Yeah I think cmake requires linking to .a files, not .la actually. So this is almost working, just need a static library for opus to link against

bastibe commented:
(And BTW, no need to rush. We're not on the clock here, nobody gets hurt if this takes time.)

joetoddsonos committed:
Use cmake to build libsndfile

joetoddsonos commented:
Ok that worked I think! Might have to do some handling of the rpath when loading in Python 

```
@rpath/libsndfile.1.dylib (compatibility version 2.0.0, current version 2.35.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
```

bastibe commented:
That's great! Thank you so very much!

Do you have a way of test-driving the new binary? I don't have access to an M1 Mac, I'm afraid. But I could of course compile an appropriate wheel to make it easier to test, if that helps.

joetoddsonos commented:
A quick test of the intel mac library built with the rpath did seem to work for me. Yeah if you could create a wheel I can get someone to test it out 👍

bastibe commented:
No problem, here you go:

[soundfile-0.12.0-py2.py3-none-macosx_11_0_arm64.whl.zip](https://github.com/bastibe/libsndfile-binaries/files/10579938/soundfile-0.12.0-py2.py3-none-macosx_11_0_arm64.whl.zip)  
[soundfile-0.12.0-py2.py3-none-macosx_10_9_x86_64.whl.zip](https://github.com/bastibe/libsndfile-binaries/files/10579947/soundfile-0.12.0-py2.py3-none-macosx_10_9_x86_64.whl.zip)

(zipped, because Github doesn't allow *.whl uploads)

bastibe commented:
By the way, no need to run a comprehensive test suite. If `soundfile.__libsndfile_version__` reports v1.2 and it can read or write mp3 files, we should be good to go.

conorsleithsonos commented:
@bastibe at the request of @joetoddsonos, I installed the arm64 wheel on my M1 Mac:
```
In [1]: from matplotlib import pyplot as plt

In [2]: import soundfile as sf

In [3]: sf.__libsndfile_version__
Out[3]: '1.2.0'

In [4]: data, sr = sf.read('test.mp3')

In [5]: plt.plot(data)
Out[5]:
[<matplotlib.lines.Line2D at 0x1126a8ee0>,
 <matplotlib.lines.Line2D at 0x1126a8f40>]

In [6]: plt.show()
```
<img width="752" alt="image" src="https://user-images.githubusercontent.com/42356132/216725895-7d22c6d9-10ab-4dc7-9832-70a3d2019cba.png">

lgtm 👍

bastibe commented:
Perfect! Thank you so very much! This is a huge help to me and the community!
